
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>HiGlass: A Multiscale Data Browser &#8212; HiGlass v1.0 documentation</title>
    <link rel="stylesheet" href="_static/higlass.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     'v1.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Tutorial" href="tutorial.html" />
  <!-- Global site tag (gtag.js) - Google Analytics -->
  <script async src="https://www.googletagmanager.com/gtag/js?id=UA-77756807-1"></script>
  <script>
    window.dataLayer = window.dataLayer || [];
    function gtag(){dataLayer.push(arguments);}
    gtag('js', new Date());

    gtag('config', 'UA-77756807-1');
  </script>
  
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />


  </head>
  <body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="higlass-a-multiscale-data-browser">
<h1>HiGlass: A Multiscale Data Browser<a class="headerlink" href="#higlass-a-multiscale-data-browser" title="Permalink to this headline">¶</a></h1>
<p><strong>develoment docs</strong></p>
<p><a class="reference external" href="higlass.io">HiGlass</a> is a viewer for large-scale genomic data. It takes
ideas introduced in genome browsers such as the <a class="reference external" href="https://genome.ucsc.edu/">UCSC Genome Browser</a> and combines them with inspirations from more
recent HiC data browsers such as <a class="reference external" href="http://www.aidenlab.org/juicebox/">Juicebox</a> and implements them in a framework
inspired by modern online maps (so-called <a class="reference external" href="http://wiki.openstreetmap.org/wiki/Slippy_Map">slippy maps</a>) to form a fast, extensible
and responsive viewer for diverse types of genomic data.</p>
<p>The HiGlass application actually consists of four distinct components:</p>
<p><a class="reference external" href="https://github.com/hms-dbmi/higlass"><strong>higlass-client</strong></a> - This is the Javascript
library that displays heatmaps, tracks, etc. In a sense, this is the meat of
the application. The vast majority of the updates are on this component.</p>
<p><a class="reference external" href="https://github.com/hms-dbmi/higlass-website"><strong>higlass-website</strong></a>  - This is the
website that embeds the higlass-client. It can contain multiple higlass clients
(as on the front page of higlass.io).</p>
<p><a class="reference external" href="https://github.com/hms-dbmi/higlass-server"><strong>higlass-server</strong></a> - The server manages data and serves it to the client in
small chunks that match the current zoom level and location. It is a python
django application that exposes an API.  For example,
<a class="reference external" href="http://higlass.io/api/v1/tilesets">http://higlass.io/api/v1/tilesets</a> lists the tilesets that the server knows
about.</p>
<p><a class="reference external" href="https://github.com/hms-dbmi/higlass-docker"><strong>higlass-docker</strong></a>  - This docker
container contains all three of the above repositories and packages them so
that they can be easily run without having to install and run each separately.
An update in any of the above repositories will be reflected in an update in
the docker container. The versions of all the components is accessible at
<a class="reference external" href="http://higlass.io/version.txt">http://higlass.io/version.txt</a></p>
<div class="section" id="track-types">
<h2>Track types<a class="headerlink" href="#track-types" title="Permalink to this headline">¶</a></h2>
<p>HiGlass supports both 1D and 2D track types that can be arranged in views to show desired
combinations of data.</p>
<div style="display: flex; align-items: center; justify-content: center; flex-wrap: wrap"><div class="figure align-center" id="id2" style="width: 100px">
<a class="reference internal image-reference" href="_images/bedlike-track-thumb.png"><img alt="_images/bedlike-track-thumb.png" src="_images/bedlike-track-thumb.png" style="height: 72px;" /></a>
<p class="caption"><span class="caption-text">Bed-like intervals</span></p>
</div>
<div class="figure align-center" id="id3" style="width: 100px">
<a class="reference internal image-reference" href="_images/gene-annotations-track-thumb.png"><img alt="_images/gene-annotations-track-thumb.png" src="_images/gene-annotations-track-thumb.png" style="height: 72px;" /></a>
<p class="caption"><span class="caption-text">Gene annotations</span></p>
</div>
<div class="figure align-center" id="id4" style="width: 100px">
<img alt="_images/chromosome-labels-thumb.png" src="_images/chromosome-labels-thumb.png" />
<p class="caption"><span class="caption-text">Chromosome labels</span></p>
</div>
<div class="figure align-center" id="id5" style="width: 100px">
<img alt="_images/line-track-thumb.png" src="_images/line-track-thumb.png" />
<p class="caption"><span class="caption-text">Line</span></p>
</div>
<div class="figure align-center" id="id6" style="width: 100px">
<img alt="_images/point-track-thumb.png" src="_images/point-track-thumb.png" />
<p class="caption"><span class="caption-text">Point</span></p>
</div>
<div class="figure align-center" id="id7" style="width: 100px">
<img alt="_images/horizontal-heatmap-thumb.png" src="_images/horizontal-heatmap-thumb.png" />
<p class="caption"><span class="caption-text">Horizontal heatmap</span></p>
</div>
<div class="figure align-center" id="id8" style="width: 100px">
<img alt="_images/horizontal-2d-rectangle-domains-thumb.png" src="_images/horizontal-2d-rectangle-domains-thumb.png" />
<p class="caption"><span class="caption-text">Horizontal rectangle domains</span></p>
</div>
<div class="figure align-center" id="id9" style="width: 100px">
<a class="reference internal image-reference" href="_images/heatmap-track-thumb.png"><img alt="_images/heatmap-track-thumb.png" src="_images/heatmap-track-thumb.png" style="height: 72px;" /></a>
<p class="caption"><span class="caption-text">Heatmap</span></p>
</div>
<div class="figure align-center" id="id10" style="width: 100px">
<a class="reference internal image-reference" href="_images/chromosome-grid-thumb.png"><img alt="_images/chromosome-grid-thumb.png" src="_images/chromosome-grid-thumb.png" style="height: 72px;" /></a>
<p class="caption"><span class="caption-text">Chromosome grid</span></p>
</div>
</div><p>It can be used to display a variety of <a class="reference external" href="track_types.html">track types</a> populated with data from
different file formats.</p>
<div class="figure align-right" id="id11" style="width: 300px">
<img alt="_images/higlass-heatmap-screenshot.png" src="_images/higlass-heatmap-screenshot.png" />
<p class="caption"><span class="caption-text">A screenshot of HiGlass displaying a Hi-C contact matrix along with tracks
for gene annotations chromosome labels.</span></p>
</div>
<div class="figure align-right" id="id12" style="width: 300px">
<img alt="_images/higlass-dual-genome-browser-screenshot.png" src="_images/higlass-dual-genome-browser-screenshot.png" />
<p class="caption"><span class="caption-text">A screenshot of HiGlass displaying a context-detail type view of two
regions of the mouse genome. The locations of the two detail regions on the
bottom are highlighted in the context view on top.</span></p>
</div>
</div>
<div class="section" id="technology">
<h2>Technology<a class="headerlink" href="#technology" title="Permalink to this headline">¶</a></h2>
<p>HiGlass is built with a variety of different open source components. The most
crucial are listed below.</p>
<ul class="simple">
<li><a class="reference external" href="https://d3js.org/">D3.js</a> - an outstanding library for online data
visualization. We make heavy use of D3’s <a class="reference external" href="https://github.com/d3/d3-scale">scales</a>, <a class="reference external" href="https://github.com/d3/d3-zoom">zoom behavior</a> and <a class="reference external" href="https://github.com/d3/d3-brush">brush behavior</a>.</li>
<li><a class="reference external" href="http://www.pixijs.com/">Pixi.js</a> - a fast, easy to use and most
importantly, well-documented library for drawing scene graphs on html
canvases. Oh, and it has WebGL support. We use Pixi to quickly render track
data and perform updates on zooming and panning.</li>
<li><a class="reference external" href="https://facebook.github.io/react/">React</a> - a widely used library for
managing state within web applications. We use React to manage the
user-interface components and lay out the positions of the views and tracks.</li>
</ul>
<div class="toctree-wrapper compound">
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="#">HiGlass</a></h1>








<ul>
<li class="toctree-l1"><a class="reference internal" href="tutorial.html">Tutorial</a><ul>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#prerequisites">Prerequisites</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#quickly-viewing-a-dataset">Quickly viewing a dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#running-higlass-locally">Running HiGlass locally</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#adding-data">Adding data</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#viewing-data-in-the-client">Viewing data in the client</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="jupyter.html">Jupyter Notebooks</a><ul>
<li class="toctree-l2"><a class="reference internal" href="jupyter.html#installation">Installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="jupyter.html#usage">Usage</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="python_api.html">Python API</a><ul>
<li class="toctree-l2"><a class="reference internal" href="python_api.html#hgflask">hgflask</a></li>
<li class="toctree-l2"><a class="reference internal" href="python_api.html#hgflask-client">hgflask.client</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="javascript_api.html">Javascript API</a><ul>
<li class="toctree-l2"><a class="reference internal" href="javascript_api.html#overview">Overview</a></li>
<li class="toctree-l2"><a class="reference internal" href="javascript_api.html#api-functions">API Functions</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="data_preparation.html">Data Preparation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="data_preparation.html#bed-files">Bed Files</a></li>
<li class="toctree-l2"><a class="reference internal" href="data_preparation.html#bedpe-like-files">Bedpe-like Files</a></li>
<li class="toctree-l2"><a class="reference internal" href="data_preparation.html#bedgraph-files">BedGraph files</a></li>
<li class="toctree-l2"><a class="reference internal" href="data_preparation.html#bigwig-files">bigWig files</a></li>
<li class="toctree-l2"><a class="reference internal" href="data_preparation.html#gene-annotation-tracks">Gene Annotation Tracks</a></li>
<li class="toctree-l2"><a class="reference internal" href="data_preparation.html#hitile-files">Hitile files</a></li>
<li class="toctree-l2"><a class="reference internal" href="data_preparation.html#cooler-files">Cooler files</a></li>
<li class="toctree-l2"><a class="reference internal" href="data_preparation.html#multivec-files">Multivec Files</a></li>
<li class="toctree-l2"><a class="reference internal" href="data_preparation.html#epilogos-data-multivec">Epilogos Data (multivec)</a></li>
<li class="toctree-l2"><a class="reference internal" href="data_preparation.html#other-data-multivec">Other Data (multivec)</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="track_types.html">Track Types</a><ul>
<li class="toctree-l2"><a class="reference internal" href="track_types.html#linking-value-scales">Linking value scales</a></li>
<li class="toctree-l2"><a class="reference internal" href="track_types.html#bed-like">Bed-like</a></li>
<li class="toctree-l2"><a class="reference internal" href="track_types.html#gene-annotations">Gene Annotations</a></li>
<li class="toctree-l2"><a class="reference internal" href="track_types.html#heatmap">Heatmap</a></li>
<li class="toctree-l2"><a class="reference internal" href="track_types.html#d-heatmap">1D Heatmap</a></li>
<li class="toctree-l2"><a class="reference internal" href="track_types.html#d-rectangle-domain">2D Rectangle Domain</a></li>
<li class="toctree-l2"><a class="reference internal" href="track_types.html#horizontal-2d-rectangle-domain">Horizontal 2D Rectangle Domain</a></li>
<li class="toctree-l2"><a class="reference internal" href="track_types.html#line">Line</a></li>
<li class="toctree-l2"><a class="reference internal" href="track_types.html#bar">Bar</a></li>
<li class="toctree-l2"><a class="reference internal" href="track_types.html#point">Point</a></li>
<li class="toctree-l2"><a class="reference internal" href="track_types.html#id5">1D Heatmap</a></li>
<li class="toctree-l2"><a class="reference internal" href="track_types.html#chromosome-labels">Chromosome Labels</a></li>
<li class="toctree-l2"><a class="reference internal" href="track_types.html#chromosome-grid">Chromosome Grid</a></li>
<li class="toctree-l2"><a class="reference internal" href="track_types.html#stacked-bars">Stacked Bars</a></li>
<li class="toctree-l2"><a class="reference internal" href="track_types.html#multiple-lines">Multiple Lines</a></li>
<li class="toctree-l2"><a class="reference internal" href="track_types.html#multiple-bar-charts">Multiple Bar Charts</a></li>
<li class="toctree-l2"><a class="reference internal" href="track_types.html#d-annotations">1D Annotations</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="developer.html">Contributor Guidelines</a></li>
<li class="toctree-l1"><a class="reference internal" href="b_viewer.html">Viewer</a><ul>
<li class="toctree-l2"><a class="reference internal" href="b_viewer.html#adding-views">Adding views</a></li>
<li class="toctree-l2"><a class="reference internal" href="b_viewer.html#closing-and-resizing-views">Closing and resizing views</a></li>
<li class="toctree-l2"><a class="reference internal" href="b_viewer.html#replacing-tracks">Replacing Tracks</a></li>
<li class="toctree-l2"><a class="reference internal" href="b_viewer.html#synchronizing-zoom-and-location">Synchronizing zoom and location</a></li>
<li class="toctree-l2"><a class="reference internal" href="b_viewer.html#adding-cross-sections">Adding cross sections</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="higlass_server.html">Server</a><ul>
<li class="toctree-l2"><a class="reference internal" href="higlass_server.html#configuration">Configuration</a></li>
<li class="toctree-l2"><a class="reference internal" href="higlass_server.html#development">Development</a></li>
<li class="toctree-l2"><a class="reference internal" href="higlass_server.html#testing">Testing</a></li>
<li class="toctree-l2"><a class="reference internal" href="higlass_server.html#api">API</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="view_config.html">View Configs</a><ul>
<li class="toctree-l2"><a class="reference internal" href="view_config.html#show-a-specific-genomic-location">Show a specific genomic location</a></li>
<li class="toctree-l2"><a class="reference internal" href="view_config.html#upload-a-viewconf-to-the-server">Upload a viewconf to the server</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="views.html">Views</a><ul>
<li class="toctree-l2"><a class="reference internal" href="views.html#genome-position-search-box">Genome Position Search Box</a></li>
<li class="toctree-l2"><a class="reference internal" href="views.html#view-synchronization">View synchronization</a></li>
</ul>
</li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="#">Documentation overview</a><ul>
      <li>Next: <a href="tutorial.html" title="next chapter">Tutorial</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2017,2018 HiGlass Authors.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.6.3</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.10</a>
      
      |
      <a href="_sources/index.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>